Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,277 @@
+using System.Linq;
+using Content.Server.Radiation.Components;
+using Content.Shared.Physics;
+using Content.Shared.Radiation.Components;
+using Content.Shared.Radiation.Events;
+using Content.Shared.Radiation.Systems;
+using Robust.Shared.Map;
+using Robust.Shared.Physics;
+using Robust.Shared.Timing;
+
+namespace Content.Server.Radiation.Systems;
+
+// main algorithm that fire radiation rays to target
+public partial class RadiationSystem
+{
+    private void UpdateGridcast()
+    {
+        // should we save debug information into rays?
+        // if there is no debug sessions connected - just ignore it
+        var saveVisitedTiles = _debugSessions.Count > 0;
+
+        var stopwatch = new Stopwatch();
+        stopwatch.Start();
+
+        var sources = EntityQuery<RadiationSourceComponent, TransformComponent>().ToArray();
+        var destinations = EntityQuery<RadiationReceiverComponent, TransformComponent>().ToArray();
+        var blockerQuery = GetEntityQuery<RadiationBlockerComponent>();
+        var resistanceQuery = GetEntityQuery<RadiationGridResistanceComponent>();
+        var transformQuery = GetEntityQuery<TransformComponent>();
+
+        // trace all rays from rad source to rad receivers
+        var rays = new List<RadiationRay>();
+        var receivedRads = new List<(RadiationReceiverComponent, float)>();
+        foreach (var (dest, destTrs) in destinations)
+        {
+            var destWorld = _transform.GetWorldPosition(destTrs, transformQuery);
+
```

Comment:
Given that there are probably going to be many more receivers than sources, the loop order should be reversed, so that the source position can be cached, rather than the the destination

Reply:
> Given that there are probably going to be many more receivers than sources, the loop order should be reversed, so that the source position can be cached, rather than the the receiver.

That was the original order, but after that I've added this code:
```csharp
foreach (var (receiver, rads) in receiversTotalRads)
{
    // update radiation value of receiver
    // if no radiation rays reached target, that will set it to 0
    receiver.CurrentRadiation = rads;

    // also send an event with combination of total rad
    if (rads > 0)
        IrradiateEntity(receiver.Owner, rads,GridcastUpdateRate);
}
```

Here I save current rad level of object or make it 0 if no ray reached receiver. This makes implementing something like geiger counter really easy. It also accumulate all `OnIrradiatedEvent` in one event. This can be easily done if I cycle in reverse order.

Reply:
Right now I update my code with one external cycle that prepares `srcWorld` and save it with components in one list:
```csharp
var sourcesData = new ValueList<(RadiationSourceComponent, TransformComponent, Vector2)>();
foreach (var (source, sourceTrs) in sources)
{
    var worldPos = _transform.GetWorldPosition(sourceTrs, transformQuery);
    var data = (source, sourceTrs, worldPos);
    sourcesData.Add(data);
}
```
This allows to save current cycle order and also calculate world position once.